From d2730f9e0cb6d26fc125035995acf8391b3a91c0 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Fri, 14 Nov 2014 19:21:45 -0500
Subject: [fs] ext4: don't count external journal blocks as overhead

Message-id: <54665649.5000206@redhat.com>
Patchwork-id: 100056
O-Subject: [RHEL7 PATCH] [fs] ext4: don't count external journal blocks as overhead
Bugzilla: 1164366
RH-Acked-by: Lukas Czerner <lczerner@redhat.com>

Bugzilla: 1164366
Upstream Status: Sent upstream, similar to commit e6d8fb3 for ext3
Build Info: Local only
Tested: Using testcase in bug, and xfstests

This was fixed for ext3 with:

e6d8fb3 ext3: Count internal journal as bsddf overhead in ext3_statfs

but was never fixed for ext4.

With a large external journal and no used disk blocks, df comes
out negative without this, as journal blocks are added to the
overhead & subtracted from used blocks unconditionally.

Signed-off-by: Eric Sandeen <sandeen@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/super.c b/fs/ext4/super.c
index f9d36be..727fea2 100644
--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -3357,8 +3357,8 @@ int ext4_calculate_overhead(struct super_block *sb)
    memset(buf, 0, PAGE_SIZE);
   cond_resched();
  }
- /* Add the journal blocks as well */
- if (sbi->s_journal)
+ /* Add the internal journal blocks as well */
+ if (sbi->s_journal && !sbi->journal_bdev)
   overhead += EXT4_NUM_B2C(sbi, sbi->s_journal->j_maxlen);
 
  sbi->s_overhead = overhead;
-- 
1.7.1